MathLib/Nonlinear/Root1D.h

```diff
@@ -0,0 +1,156 @@
+#include <cmath>
+#include <type_traits>
+
+namespace MathLib
+{
+namespace Nonlinear
+{
+
+namespace detail
+{
+
+//! Tells if \c a and \c b have the same sign.
+bool same_sign(double a, double b)
+{
+    // note: signbit() casts integers to double and distinguishes between +0.0 and
+    // -0.0. However, the latter is not a problem, because if f(x0) == 0, we've
+    // already found the root x0 that we are searching for.
+    return std::signbit(a) == std::signbit(b);
+}
+
+}
+
+
+//! Use the regula falsi method to find the root of some scalar function of one
+//! variable.
+template<typename SubType, typename Function>
+class RegulaFalsi
+{
+public:
+    //! Initializes finding a root of the \c Function \c f in the interval
+    //! [\c a, \c b].
+    RegulaFalsi(Function const& f, double a, double b)
+        : _f(f), _a(a), _b(b), _fa(f(a)), _fb(f(b))
+    {
+        static_assert(
+                std::is_same<double, decltype(f(0.0))>::value,
+                "Using this class for functions that do not return double"
+                " involves a lot of casts. Hence it is disabled.");
+
+        if (_fa == 0.0) {
+            _b = _a;
+        } else if (_fb == 0.0) {
+            _a = _b;
+        } else {
+            assert(!detail::same_sign(_fa, _fb));
+        }
+    }
+
+    //! Do \c num_steps iteration of regula falsi.
+    void step(const unsigned num_steps)
+    {
+        for (unsigned i=0; i<num_steps; ++i)
+        {
+            if (_a == _b) return;
+
+            const double s = (_fb - _fa)/(_b - _a);
+            const double c = _a - _fa/s;
+            const double fc = _f(c);
+
+            if (fc == 0.0) {
+                _a = _b = c;
+                return;
+            } else if (!detail::same_sign(fc, _fb)) {
+                _a = _b; _fa = _fb;
+                _b =  c; _fb =  fc;
+            } else {
+                const double m = SubType::get_m(_fa, _fb, fc);
+                _fa *= m;
+                _b = c;
+                _fb = fc;
+            }
+        }
+    }
+
+    //! Returns the current estimate of the root.
+    double getResult() const
+    {
+        if (_a == _b) return _a;
+
+        const double s = (_fb - _fa)/(_b - _a);
+        const double c = _a - _fa/s;
+
+        return c;
+    }
+
+    //! Returns the size of the current search interval.
+    double getRange() const { return std::fabs(_a - _b); }
+
+private:
+    Function const& _f;
+    double _a, _b, _fa, _fb;
+};
+
+
+/*! Creates a new instance of \c RegulaFalsi that uses the given modified iteration
+ *  method \c SubType.
+ *
+ * \see https://en.wikipedia.org/wiki/False_position_method#Improvements_in_regula_falsi
+ */
+template<typename SubType, typename Function>
+RegulaFalsi<SubType, Function>
+makeRegulaFalsi(Function const& f, double const a, double const b)
+{
+    return RegulaFalsi<SubType, Function>(f, a, b);
+}
+
+
+//! Used by \c RegulaFalsi in the original regula falsi algorithm.
+struct Unmodified
```

Comment:
Maybe instead of `Nonlinear`? The types are already in a pretty nested namespace. Furthermore the other members of MathLib/Nonlinear look pretty obsolete.